articles/055_ros_parameter_design.md

```diff
@@ -0,0 +1,200 @@
+
+- Gonzalo's research on parameters.
+  - Discussion: https://groups.google.com/forum/#!topic/ros-sig-ng-ros/YzCmoIsN0o8 and https://groups.google.com/forum/#!searchin/ros-sig-ng-ros/parameter/ros-sig-ng-ros/fwDBcei5Ths/L6ORPfjUDXYJ
+  - Protoype: https://github.com/abellagonzalo/dynamic_config
+  - Final Notes:  [http://wiki.ros.org/sig/NextGenerationROS/Parameters](http://wiki.ros.org/sig/NextGenerationROS/Parameters)
+- Thibault's nodeparam draft REP: https://github.com/tkruse/rep/blob/nodeparam/nodeparam-REP.rst
+
+## Ideal System
+
+It is useful to consider the ideal system to understand how it relates to the current system and how a new system could work.
+We'd like to support both the use cases of ROS 1.0 built in parameters as well as dynamic parameters.
+An ideal parameter system would have the qualities laid out in the following paragraphs.
+
+### Set parameter values
+
+At the most basic, a parameter system must be able to accept the setting of parameter values.
+An extension is to accept a group of parameters updated atomically.
+This allows you to update coupled parameters such as PID gains without worring about invalid intermediate states.
+
+### Get parameter values
+
+Also at the core functionality return the value of one or more parameters which have previously been set.
+If the value of a parameter is requested which has not been set it will explicitly return `unset`.
+Parameters can be queried individually or in groups atomically so as to get a consistent state.
+
+### Unset a parameter value
+
+If a parameter has been set it should be able to be unset.
+
+### Introspection of available parameters
+
+Provide a list of currently set parameters.
+
+### Notifications of change
+
+When a parameter changes value it should be possible to get a notification.
+
+### Reject parameter changes
+
+If a parameter change is requested to an invalid value, the parameter server should be able to reject the change.
+Related to this, it should also be able to communicate the preconditions which are necessary but not sufficient to determine the acceptance of a change.
+
+### Introspection of expected parameters
+
+When using parameters, providing a list of expected parameters can prevent accidentally setting parameters with typos in their names etc.
+
+### Control parameter lifetime
+
+When parameters are set, there should be an understanding of what the lifetime of the parameter value will be and what conditions will clear its value.
+
+### Unambigious naming
+
+All parameters should be unambiguously addressable.
+
+For example, one of the challenges of the current system is that there is a naming ambiguity between nodes and parameters `/foo/bar/baz` could be a node `/foo/bar/baz` or a private parameter `baz` on node `/foo/bar`.
+
+### Logging
+
+All changes to parameters should be visible when logging.
+And when playing back a log file the parameter changes should be able to be reapplied.
```

Comment:
I'm not sure about this line, since it's possible that any change that is played back could be rejected. Maybe it would be more appropriately scoped to say that it should be possible to playback parameter events.